### ISSUE TITLE
Order of non-standard vendor-prefixed properties within declaration-block-properties-order

### ISSUE THREAD

**Issue by igoradamenko** [stylelint/stylelint#1751]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

A bug with a rule `-webkit-overflow-scrolling`. If I use `declaration-block-properties-order` linter want to put it after `overflow`. Quote from [doc](http://stylelint.io/user-guide/rules/declaration-block-properties-order/#options):

"If an (**unprefixed**) property name is not included in your array and it contains a hyphen (e.g. padding-left), this rule will look for the string before that first hyphen in your array (e.g. padding) and use that position."

> Which rule, if any, is this issue related to?

`-webkit-overflow-scrolling`.

> What CSS is needed to reproduce this issue?

```css
div {
	overflow: auto;    
	overflow-x: auto;
	width: 100%;
	-webkit-overflow-scrolling: touch;
}
```

> What stylelint configuration is needed to reproduce this issue?

```json
{
  "rules": {
    "declaration-block-properties-order": [
      "overflow",
      "width"
    ]
  }
}
```

> Which version of stylelint are you using?

`7.0.3`

> How are you running stylelint: CLI, PostCSS plugin, Node API?

Using Gulp plugin.

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

Don't think so. But I use SCSS.

> What did you expect to happen?

No warnings to be flagged.

> What actually happened (e.g. what warnings or errors you are getting)?

The following warnings were flagged:

```shell
 369:2   ×  Expected                     declaration-block-properties-order
            "-webkit-overflow-
            scrolling" to come before
            "width"
```

**Comment by jeddy3** [comment]
@igoradamenko Thanks for using the issue template and for the detailed report.

I think the current behaviour, i.e. expecting `-webkit-overflow-scrolling` to be grouped with the other `overflow-*` properties, makes the most sense. I think the documentation can improved to explain this behaviour more clearly though, as at the moment the use of the word "unprefixed" in brackets in a little ambiguous. How's about?:

"If a property name is not included in your array and it contains a hyphen (e.g. padding-left), this rule will look for the string before that first hyphen in your array (e.g. padding) and use that position. Any prefixes are removed from the property names *before* their inclusion in the array is checked."

**Comment by igoradamenko** [comment]
@jeddy3 It sounds good for me. 

But what if I want to place `-webkit-overflow-scrolling` in another place? For instance to the bottom of properties block. If I understand right (and right now it works that way) I should use unprefixed property names in `declaration-block-properties-order` array. If so I think it would be good to add one more example or some description of this situation to the documentation. I'm not sure that it is described there right now.

**Comment by jeddy3** [comment]
> For instance to the bottom of properties block

You'd put `overflow-scrolling` as the last item in the array. We can clarify the docs around that too as part of this issue.

**Comment by igoradamenko** [comment]
> We can clarify the docs around that too as part of this issue.

Yeah, I meant exactly that. Let's do it! :koala:

**Comment by woble** [comment]
I do believe stylelint should handle this in a similar fashion as [`property-no-vendor-prefix`](http://stylelint.io/user-guide/rules/property-no-vendor-prefix/) rule does because there's no such property as `overflow-scrolling`. `-webkit-overflow-scrolling` is a non-standard vendor-prefixed property.

**Comment by jeddy3** [comment]
> I do believe stylelint should handle this in a similar fashion as [`property-no-vendor-prefix`](http://stylelint.io/user-guide/rules/property-no-vendor-prefix/) rule does because there's no such property as `overflow-scrolling`.

As people might already be relying on this behaviour, i.e. using something like `overflow-scrolling` in their array, and that vendor prefixes are supposedly going away in favour of browser flags, I think we just document this idiosyncrasy and move on. Anyone have any objections to this?

**Comment by woble** [comment]
> [...] and that vendor prefixes are supposedly going away in favour of browser flags, [...]

Can you provide some source for that? I know that Chrome uses browser flags to enable some exprimental features, but there are still plenty of prefixed proprietary properties in browsers like Firefox and Safari that will be there indefinitely.

> As people might already be relying on this behaviour [...]

How does this apply when rules are being deprecated and/or renamed? People have to change their behaviour accordingly, don't they?

I'm happy with either way of how you guys decide to solve this issue, it's just consistency is key here, i.e. why should one rule respect non-standard vendor prefixed props and the other one shouldn't.

**Comment by davidtheclark** [comment]
> I do believe stylelint should handle this in a similar fashion as property-no-vendor-prefix rule does because there's no such property as overflow-scrolling.
@woble Can you be specific about what you'd like the rule to do?

Meanwhile, I'm going to label this as a documentation issue.

**Comment by woble** [comment]
My understanding of the current behaviour, correct me if I am wrong, is that the rule ...
1. ... ignores vendor prefixed properties and makes own decision of where that property should be, and ...
2. ... makes assumption that for example if `mask` is specified in the rule and `mask-image` is not, `mask-image` should come after `mask` when declared. Same incorrect behaviour applies to when a random non-existent property like `mask-foo` is declared, stylelint assumes it should be in the vicinity of `mask`.

While it makes sense for things like `filter` and `-webkit-filter`, it does not make sense in case of vendor prefixed proprietary properties that do no have a standard counterpart, or nonsensical properties that are derived from an existing property as shown in 2.

That's why I said that I believe that the rule should probably behave the same way as `property-no-vendor-prefix` — checking the property against `caniuse` with `autoprefixer`. Alternatively the rule should be completely agnostic and simply follow what is specified in the rule list and do not make any assumptions that are not desired by the user. In either case, a rule option can be implemented that disables current behaviour. This avoids users, who do rely on current behaviour (as mentioned by @jeddy3), having to make changes to the their rule list and/or CSS.

**Comment by davidtheclark** [comment]
@woble Thanks for the explanation, but I think it would be made significantly more concrete and clear if you provide some example code with expected warnings. I'm still not totally sure what you *do* want the rule to do.

**Comment by woble** [comment]
I've read the documentation on the rule again and it seems I might have been somewhat incorrect in regards to properties with hyphens, i.e. `mask-image`; it makes sense to have them in relative position to the parent, i.e. `mask`. Sorry for any confusion in my last post.

I still do believe that prefixed vendor proprietary properties, i.e. `-webkit-overflow-scrolling`, should be handled differently, as described by the OP.
